File: android/window/src/main/java/org/dweb_browser/window/core/windowInstancesManager.kt
package org.dweb_browser.window.core

import org.dweb_browser.helper.ChangeableMap
import org.dweb_browser.window.core.constant.UUID

/**
 * 窗口实例管理
 */
class WindowInstancesManager {
  /**
   * 所有的窗口实例
   */
  private val instances = ChangeableMap<UUID, WindowController>()
  fun get(id: UUID) = instances[id]
  fun add(window: WindowController) {
    instances[window.id] = window
    window.onClose {
      instances.remove(window.id)
    }
  }

  val onAdd = instances.onChange
  val onRemove = instances.onChange
}

val windowInstancesManager = WindowInstancesManager()

File: next/dweb-browser/src/Platforms/iOS/MicroModule/Core/WindowInstancesManager.cs
namespace DwebBrowser.MicroService.Core;

public class WindowInstancesManager
{
    public static WindowInstancesManager WindowInstancesManagerInstance = new();

    /// <summary>
    /// 所有的窗口实例
    /// </summary>
    public ChangeableMap<UUID, WindowController> Instances = new();

    public WindowController? Get(UUID id) => Instances.GetValueOrDefault(id);

    public void Add(WindowController window)
    {
        Instances.TryAdd(window.Id, window);
        window.OnClose.OnListener += async (_, _) =>
        {
            Instances.Remove(window.Id);
        };
    }
}

